libyul/backends/wasm/BinaryTransform.cpp

```diff
@@ -556,16 +607,22 @@ bytes BinaryTransform::memorySection()
 	return makeSection(Section::MEMORY, std::move(result));
 }
 
-bytes BinaryTransform::globalSection()
+bytes BinaryTransform::globalSection(vector<GlobalVariableDeclaration> const& _globals)
 {
+	yulAssert(m_globals.size() == _globals.size(), "");
+	// ASSUMPTION: _globals[i] and m_globals[i] are the same variable
+
 	bytes result = lebEncode(m_globals.size());
-	for (size_t i = 0; i < m_globals.size(); ++i)
+	for (wasm::GlobalVariableDeclaration const& global: _globals)
```

Comment:
I don't see a point for
```
	for (size_t i = 0; i < _module.globals.size(); ++i)		for (size_t i = 0; i < _module.globals.size(); ++i)
		bt.m_globals[_module.globals[i].variableName] = i;			bt.m_globals[_module.globals[i].variableName] = i;
```

if we here use both `_globals` and `m_globals`. Maybe I am missing the bigger picture or just this could be refactored.

Reply:
It's initialization so it needs access to both. `m_globals` is used later during tree traversal to make global variable lookups a bit more efficient (it's a map and `_module.globals` is a list and would require linear search).

But you have a point that still requiring access to `_module.globals` after the initialization is not a great design. We would not need to refer back to it if `m_globals` could store variable types too. I just followed the existing design that's used for `m_functions`/`_functions` and `m_functions`/`_imports`. I'll refactor it in a separate PR.

Reply:
@axic I tried refactoring it a bit. Please take a look at #9151.